core/common/src/main/java/alluxio/resource/DynamicResourcePool.java

```diff
@@ -0,0 +1,486 @@
+        throw new IllegalArgumentException(
+            "Resource " + resource.toString() + " was not acquired from this resource pool.");
+      }
+      ResourceInternal<T> resourceInternal = mResources.get(resource);
+      resourceInternal.setLastAccessTimeMs(System.currentTimeMillis());
+      mResourceAvailable.add(resourceInternal);
+      mNotEmpty.signal();
+    } finally {
+      mLock.unlock();
+    }
+  }
+
+  /**
+   * Closes the pool and clears all the resources.
+   */
+  @Override
+  public void close() {
+    try {
+      mLock.lock();
+      if (mResourceAvailable.size() != mResources.size()) {
+        LOG.warn("Some resources are not released when closing the resource pool.");
+      }
+      for (ResourceInternal<T> resourceInternal : mResourceAvailable) {
+        closeResourceSync(resourceInternal.mResource);
+      }
+    } finally {
+      mLock.unlock();
+    }
+  }
+
+  @Override
+  public int size() {
+    try {
+      mLock.lock();
+      return mResources.size();
+    } finally {
+      mLock.unlock();
+    }
+  }
+
+  /**
+   * @return true if the pool is full
+   */
+  private boolean isFull() {
+    boolean full = false;
+    try {
+      mLock.lock();
+      full = mResources.size() >= mMaxCapacity;
+    } finally {
+      mLock.unlock();
+    }
+    return full;
+  }
+
+  /**
+   * Adds a newly created resource to the pool. The resource is not available when it is added.
+   *
+   * @param resource
+   * @return true if the resource is successfully added
+   */
+  private boolean add(ResourceInternal<T> resource) {
+    try {
+      mLock.lock();
+      if (mResources.size() >= mMaxCapacity) {
+        return false;
+      } else {
+        mResources.put(resource.mResource, resource);
+        return true;
+      }
+    } finally {
+      mLock.unlock();
+    }
+  }
+
+  /**
+   * Removes an existing resource from the pool.
+   *
+   * @param resource
+   */
+  private void remove(T resource) {
+    try {
+      mLock.lock();
+      mResources.remove(resource);
+    } finally {
+      mLock.unlock();
+    }
+  }
+
+  /**
+   * @return the most recently used resource
+   */
+  private ResourceInternal<T> poll() {
+    ResourceInternal<T> resource;
+    try {
+      mLock.lock();
+      resource = mResourceAvailable.pollFirst();
+    } finally {
+      mLock.unlock();
+    }
+    return resource;
+  }
+
+  /**
+   * Check whether the resource is healthy. If not retry.
+   *
+   * @param resource the resource to check
+   * @param endTimeMs the end time to wait till
+   * @return the resource
+   * @throws IOException if it fails to create a resource
+   * @throws TimeoutException if it times out to wait for a resource
+   */
+  private T checkHealthyAndRetry(T resource, long endTimeMs) throws IOException, TimeoutException {
+    if (isHealthy(resource)) {
+      return resource;
+    } else {
+      LOG.info("Clearing unhealthy resource {}.", resource);
+      closeResource(resource);
+      remove(resource);
```

Comment:
done. it doesn't matter here because no one can acquire it anyway since it is not the available set.